Replace google-map with openStreetMap in map interaction.

DubeySandeep opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Replace google-map with OpenStreetMap in map interaction.
Also, fixes #1269.

**TODO**
- [x] Testing changes in mobile.
- [x] Cleanup!
- [x] Ready for review.
<!--
## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [ ] The PR is made from a branch that's **not** called "develop".
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.
  -->

DubeySandeep committed:
Replaced google-map with openStreetMap in interaction.

nithusha21 commented:
Hi @DubeySandeep , could you please add a changelog label (before merging) as mentioned [here](https://groups.google.com/forum/#!topic/oppia-dev/4p555mIskWE). Thanks!

seanlip commented on extensions/dependencies/ui_leaflet.html:
```diff
@@ -0,0 +1,4 @@
+<link rel="stylesheet" href="/third_party/static/leaflet-1.4.0/leaflet.css">
+<script src="/third_party/static/leaflet-1.4.0/leaflet.js"></script>
+<script src="/third_party/static/angular-ui-leaflet-1.0.3/ui-leaflet.min.no-header.js"></script>
+<script src="/third_party/static/angular-simple-logger-1.0.3/angular-simple-logger.min.js"></script>
```
Why do we need "simple logger"? Can we do without it?

DubeySandeep replied:
This is needed as a dependency for ui-leaflet (I have added a comment above this line.)

seanlip commented on extensions/interactions/InteractiveMap/directives/OppiaInteractiveInteractiveMapDirective.js:
```diff
@@ -52,6 +54,20 @@ oppia.directive('oppiaInteractiveInteractiveMap', [
               opacity: '0.8'
             };
           };
+          var ICON = {
```
Seems like a weird constant name. Maybe just inline it in addNewMarker() since it's only used once? You can precompute the URLs beforehand and store them as top-level variables if you like.

DubeySandeep replied:
I wasn't sure about the name earlier, I wanted to move it to a oppia.constant so that I can use it in other places but it uses UrlInterpolationService so not possible). I've moved this to the dict as suggested.

seanlip commented on extensions/interactions/InteractiveMap/directives/OppiaInteractiveInteractiveMapDirective.js:
```diff
@@ -72,61 +96,44 @@ oppia.directive('oppiaInteractiveInteractiveMap', [
           };
 
-          $scope.onMouseOver = function() {
+          $scope.$on('leafletDirectiveMap.mouseover', function(event) {
```
Drop "event" if it's not being used, ditto below.

DubeySandeep replied:
Done!

seanlip commented on extensions/interactions/InteractiveMap/directives/OppiaResponseInteractiveMapDirective.js:
```diff
@@ -34,10 +34,12 @@ oppia.directive('oppiaResponseInteractiveMap', [
-          latLongPair + '&sensor=false';
+          'https://staticmap.openstreetmap.de/staticmap.php?center=' +
+          latLongPair + '&zoom=14&size=865x512&maptype=mapnik';
+        // 'https://maps.googleapis.com/maps/api/staticmap?' +
```
Delete?

DubeySandeep replied:
Done!

seanlip commented on extensions/objects/templates/CoordTwoDimEditor.js:
```diff
@@ -17,81 +17,54 @@ oppia.directive('coordTwoDimEditor', [
-                map: $scope.map,
-                position: latLng
-              });
+        angular.extend($scope, {
```
Write $scope.center = ..., $scope.markers = ..., $scope.events = ...

DubeySandeep replied:
I was too thinking of doing this. Thanks for the suggestion. Done!

seanlip commented on extensions/objects/templates/CoordTwoDimEditor.js:
```diff
@@ -17,81 +17,54 @@ oppia.directive('coordTwoDimEditor', [
+            zoom: 0
+          },
+          markers: {
+            markersmarkers: {
```
Just checking, is this a typo ("markersmarkers")? Seems like a weird key name.

DubeySandeep replied:
I had found this somewhere in examples, although the correct name is mainMarker but it was working with this also, don't know how ui-leaflet handle this. Changed!

seanlip commented on manifest.json:
```diff
@@ -272,6 +272,30 @@
+        "url": "https://unpkg.com/leaflet@1.4.0/dist",
+        "rootDirPrefix": "leaflet-",
+        "targetDirPrefix": "leaflet-",
+        "files": ["leaflet.js","leaflet.css"]
```
add space after comma

DubeySandeep replied:
Done!

seanlip reviewed (changes requested):
Just did an initial pass and left a few notes, thanks @DubeySandeep! Overall the "shape" of the changes looks good.

seanlip commented:
Does this fix #1269 btw?

DubeySandeep committed:
Changes made after testing UI on mobile devices.

DubeySandeep committed:
Simple UI changes.

DubeySandeep commented:
Hi @seanlip, I had created this draft PR for testing UI behavior it wasn't in a clean state earlier, I have cleaned it (thanks for the last review!) and tested this with Android device (working fine). Now it's open for review. 
re > Does this fix #1269 btw?

Yep, it resolves that issue, I've added this to the explanation section.

DubeySandeep commented:
@nithusha21, I'm not sure about the label for this PR, I was expecting to have a separate label for interactions (as we treat interactions as an extension)  but didn't find one so added "[ Improvements to editors and players](/oppia/oppia/labels/CHANGELOG%3A%20Improvements%20to%20editors%20and%20players)".

DubeySandeep committed:
Updated correct depen. URL.

seanlip commented on extensions/objects/templates/coord_two_dim_editor_directive.html:
```diff
@@ -1,7 +1,15 @@
+  <leaflet id="coordTwoDimEditor" lf-center="mapCenter" event-broadcast="mapEvents" markers="mapMarkers"></leaflet>
 </div>
+<style>
+  #coordTwoDimEditor {
```
Could we use a CSS class instead?

DubeySandeep replied:
Internally ui-leaflet use to access this id style and apply this to map container. Changing this to the class is creating some stylistic issue.

seanlip replied:
See comment above re generating the ID and handling the CSS style separately -- is that possible? (If not, no worries, just wanted to check.)

Also for any case where you use IDs for CSS, make sure to add a comment explaining what's going on.

DubeySandeep committed:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-google-map

DubeySandeep committed:
Fixed state navigation issue.

DubeySandeep committed:
Fixed lint issue.

apb7 commented:
Hi @DubeySandeep, please bring your PR branch up-to-date with oppia/develop. Thanks!

DubeySandeep committed:
Added review changes.

DubeySandeep committed:
Added review changes.

DubeySandeep commented:
@seanlip, I've added the changes PTAL!

Also, I've created an issue (#6354) for switching tab changes map center in preview mode (for editors page.).

DubeySandeep committed:
Made some stylistic changes after following a user journey.

DubeySandeep commented:
@kevinlee12, @vojtechjelinek, @ankita240796, @aks681 can you please take a look? (this need to be merged before March release cut!)

DubeySandeep commented:
@apb7, I'm not sure why the lint tests are failing in CircleCi but passes locally and in Travis,  any suggestion?

ankita240796 reviewed (changes requested):
Hi @DubeySandeep 

Can you also remove type definitions for google map and add type definitions for openStreetMap. The changes will be made in [install_third_party.sh](https://github.com/oppia/oppia/blob/develop/scripts/install_third_party.sh#L46) and [tsconfig.json](https://github.com/oppia/oppia/blob/develop/tsconfig.json#L5).

Do let me know if you face any issue, Thanks!

DubeySandeep committed:
Remove googlemaps from node_modules.

DubeySandeep commented:
@ankita240796, Thanks for pointing this, I've removed googleMap but I don't think we need to add type for openStreetMap. (Let me know if I have misunderstood something.)

ankita240796 commented:
Hi @DubeySandeep, Thanks for removing typings for googlemaps. I meant to say adding type definitions for leaflet from [here](https://www.npmjs.com/package/@types/leaflet).

DubeySandeep committed:
Added leaflet type for npm.

DubeySandeep commented:
@ankita240796, I've made the changes PTAL!

DubeySandeep committed:
made changes in tsconfig.json

DubeySandeep commented:
@ankita240796, I've made the changes PTAL!

ankita240796 reviewed (approved):
LGTM, Thanks @DubeySandeep!

seanlip commented on extensions/interactions/InteractiveMap/directives/interactive_map_interaction_directive.html:
```diff
@@ -1,13 +1,5 @@
-  }
-
-  #map_canvas {
+  #interactiveMap {
```
I worry slightly about this. Should we use a CSS class for the style, and additionally generate a random ID so that each map instance is unique? I'm not sure.

DubeySandeep replied:
Added class. (Done!)

seanlip commented on extensions/objects/templates/CoordTwoDimEditor.js:
```diff
@@ -16,82 +16,67 @@ oppia.directive('coordTwoDimEditor', [
-          }
-        }, 100);
+        $scope.$on('leafletDirectiveMap.coordTwoDimEditor.click',
+          function(event, args) {
```
Typically we use evt rather than event (here and below)

DubeySandeep replied:
Done!

apb7 commented:
> @apb7, I'm not sure why the lint tests are failing in CircleCi but passes locally and in Travis, any suggestion?

@DubeySandeep, Can you please update your branch once again, with upstream develop? We are not running lint tests on CircleCI atm. Let me know if there's any problem. Thanks!

DubeySandeep committed:
Merge branch 'develop' of https://github.com/oppia/oppia into remove-google-map

DubeySandeep committed:
Added CSS class.

DubeySandeep commented:
Adding style property to class worked out but had to change the width and height for mobile view a little.

![image](https://user-images.githubusercontent.com/16653571/53657822-46ec9880-3c7d-11e9-94f0-eab4b7b333d4.png)

DubeySandeep commented:
@apb7, Thanks for the help, I've updated the branch now! Can you please do a final review and approve this PR?

seanlip commented:
Thanks @DubeySandeep! Approving from my perspective. The mobile view looks good!

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks for using open source maps!

kevinlee12 reviewed (approved):
lgtm from code owners perspective!

aks681 reviewed (approved):
LGTM!

apb7 reviewed (approved):
LGTM from code owner's perspective. Thanks!
